<div class="paragraph">
<p>In Python 3, attempting to catch in an <code>except</code> statement an object which does not derive from BaseException will raise a <code>TypeError</code>. In Python 2 it is possible to raise old-style classes but this shouldn&#8217;t be done anymore in order to be compatible with Python 3.</p>
</div>
<div class="paragraph">
<p>In order to catch multiple exceptions in an <code>except</code> statement, a tuple of exception classes should be provided.</p>
</div>
<div class="paragraph">
<p>If you are about to create a custom Exception class, note that custom exceptions should inherit from <code>Exception</code>, not <code>BaseException</code>. <code>Exception</code> allows people to catch all exceptions except the ones explicitly asking the interpreter to stop, such as <code>KeyboardInterrupt</code> and <a href="https://docs.python.org/3/library/exceptions.html#GeneratorExit"><code>GeneratorExit</code></a> which is not an error. See <a href="https://www.python.org/dev/peps/pep-0352/#exception-hierarchy-changes">PEP 352</a> for more information.</p>
</div>
<div class="paragraph">
<p>This rule raises an issue when the expression used in an <code>except</code> statement is not a class deriving from <code>BaseException</code> nor a tuple of such classes.</p>
</div>
<div class="sect1">
<h2 id="_noncompliant_code_example">Noncompliant Code Example</h2>
<div class="sectionbody">
<div class="listingblock">
<div class="content">
<pre>class CustomException:
    """An Invalid exception class."""

try:
    "a string" * 42
except CustomException:  # Noncompliant
    print("exception")
except (None, list()):  # Noncompliant * 2
    print("exception")

try:
    "a string" * 42
except [TypeError, ValueError]:  # Noncompliant. Lists are not accepted.
    print("exception")
except {TypeError, ValueError}:  # Noncompliant. Sets are not accepted.
    print("exception")</pre>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_compliant_solution">Compliant Solution</h2>
<div class="sectionbody">
<div class="listingblock">
<div class="content">
<pre>class MyError(Exception):
    pass

try:
    "a string" * 42
except (MyError, TypeError):
    print("exception")</pre>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_see">See</h2>
<div class="sectionbody">
<div class="ulist">
<ul>
<li>
<p>Python documentation - <a href="https://docs.python.org/3/tutorial/errors.html">Errors and Exceptions</a></p>
</li>
<li>
<p>Python documentation - <a href="https://docs.python.org/3/reference/compound_stmts.html#except">the <code>try</code> statement</a></p>
</li>
<li>
<p><a href="https://www.python.org/dev/peps/pep-0352/#exception-hierarchy-changes">PEP 352 - Required Superclass for Exceptions</a></p>
</li>
</ul>
</div>
<hr/>
</div>
</div>
<div class="sect1">
<h2 id="_implementation_specification">Implementation Specification</h2>
<div class="sectionbody">
<div class="paragraph">
<p>(visible only on this page)</p>
</div>
<div class="sect2">
<h3 id="_message">Message</h3>
<div class="ulist">
<ul>
<li>
<p>Change this expression to be a class deriving from BaseException or a tuple of such classes.</p>
</li>
</ul>
</div>
</div>
<div class="sect2">
<h3 id="_highlighting">Highlighting</h3>
<div class="paragraph">
<p>The expression which does not resolve in a valid exception class or a tuple of such classes.</p>
</div>
<hr/>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_comments_and_links">Comments And Links</h2>
<div class="sectionbody">
<div class="paragraph">
<p>(visible only on this page)</p>
</div>
<div class="sect2">
<h3 id="_relates_to_s5632">relates to: S5632</h3>

</div>
<div class="sect2">
<h3 id="_on_5_mar_2020_141755_nicolas_harraudeau_wrote">on 5 Mar 2020, 14:17:55 Nicolas Harraudeau wrote:</h3>
<div class="paragraph">
<p>Even if it is possible to catch a class which is not a BaseException in python 2, this rule is still a bug for the following reasons:</p>
</div>
<div class="ulist">
<ul>
<li>
<p>BaseException exists in Python 2 too. There is no reason not to use it. It is advised to use exceptions derived from BaseException since python 2.5 released in 2006.</p>
</li>
<li>
<p>Python 2 is not maintained anymore and there are tools enabling developpers to migrate from python 2 to python 3. When developers migrate this will become a bug.</p>
</li>
<li>
<p>We already have RSPEC-5632 as a Bug and nobody complains. Letâ€™s remain consistent. If we change the rule to code smell we have to change this other rule too.</p>
</li>
<li>
<p>old style classes are the only exception in python 2. Trying to raise a string or a list will fail in in both python 2 and python 3. This exception would be a corner case.</p>
</li>
</ul>
</div>
</div>
</div>
</div>